### PR TITLE
Implement metric scaling

### PR DESCRIPTION
**Pull request by AdrienCorenflos**
# Thank you for opening a PR!

This is a change that relates to the PR https://github.com/blackjax-devs/blackjax/pull/731.
The idea is to allow metrics to scale a vector (possibly momentum, possibly otherwise) by the cholesky of the mass matrix or its inverse.

DRAFT PR so far, taking comments on interface and tentative implementation. Nothing is tested yet.

A comment: for efficiency reasons (in the context of Riemannian metrics), I have allowed the scale function to take a *sequence* of elements to rescale, so as to avoid recomputing the mass matrix all the time. There may be a cleaner way to do this but I can't see it.

### CODE REVIEW

**Comment by AdrienCorenflos** [comment]
@ismael-mendoza this is what you need right?

File: blackjax/mcmc/metrics.py
Comment (junpenglao) [inline blackjax/mcmc/metrics.py:331]:
```suggestion
    return Metric(momentum_generator, kinetic_energy, is_turning, scale)
```

**Comment by junpenglao** [comment]
> A comment: for efficiency reasons (in the context of Riemannian metrics), I have allowed the scale function to take a sequence of elements to rescale, so as to avoid recomputing the mass matrix all the time. There may be a cleaner way to do this but I can't see it.

I see... I dont love it but I also dont hate it LOL

**Comment by ismael-mendoza** [comment]
Thanks @AdrienCorenflos I think this will work great, happy to incoporate this into #731 once it's merged!

**Comment by AdrienCorenflos** [comment]
I am a bit unhappy with one design choice: for the function scale to *mean* the same thing for all metrics I had to treat inputs differently: this comes from the fact that euclidean metrics take an inverse mass_matrix, while Riemannian take the mass matrix... I'm not sure what the rationale was behind this, but we will need to rethink it down the line. Maybe Riemannian should also take the inverse mass matrix? Either way, not a problem for this PR.

**Comment by junpenglao** [comment]
A likely over-engineered solution here is to create a new covariance matrix class that has both original and inverse form, kind of like what TFP does for Gaussian Process https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/math/psd_kernels/positive_semidefinite_kernel.py

**Comment by AdrienCorenflos** [comment]
I changed the api a bit -> you now only give scale(position, vector, inv) rather than something complicated. This is because I remembered we are using JAX and I shouldn't reinvent the wheel: if people want several scalings they should use vmap and that's it.

**Comment by AdrienCorenflos** [comment]
@junpenglao any comments or go?

**Comment by junpenglao** [comment]
Let me take another look tomorrow

File: blackjax/mcmc/metrics.py
Comment (junpenglao) [inline blackjax/mcmc/metrics.py:36]: Let's add the type into https://github.com/blackjax-devs/blackjax/blob/main/blackjax/types.py instead, copying the 2 lines from chex

https://github.com/google-deepmind/chex/blob/343d03af64d0ead945fdda194b9d5c5111eb31d7/chex/_src/pytypes.py#L53-L54

reason: chex usage should be test only.

File: blackjax/mcmc/metrics.py
Comment (junpenglao) [inline blackjax/mcmc/metrics.py:68]: turn in to a Protocal as well? That way we can see also the names

File: blackjax/mcmc/metrics.py
Comment (junpenglao) [inline blackjax/mcmc/metrics.py:184]:
```suggestion
        position: ArrayLikeTree, element: ArrayLikeTree, inv: bool
```

Same below

File: blackjax/mcmc/metrics.py
Comment (junpenglao) [inline blackjax/mcmc/metrics.py:296]: Should always be false for Reinmannian kinetic energy for now?

Comment (AdrienCorenflos) [reply blackjax/mcmc/metrics.py:296]: No it's a different inv -> the is_inv above is about wether the mass matrix was given in inverse form, this one is about wether your want to scale the vector by the inverse of the cholesky or the cholesky.

File: blackjax/mcmc/metrics.py
Comment (AdrienCorenflos) [reply blackjax/mcmc/metrics.py:184]: Below where please?

File: blackjax/mcmc/metrics.py
Comment (AdrienCorenflos) [reply blackjax/mcmc/metrics.py:68]: Ok

File: blackjax/mcmc/metrics.py
Comment (AdrienCorenflos) [reply blackjax/mcmc/metrics.py:36]: Ok

File: blackjax/mcmc/metrics.py
Comment (junpenglao) [inline blackjax/mcmc/metrics.py:276]:
```suggestion
        position: ArrayLikeTree, element: ArrayLikeTree, inv: bool
```

**Comment by AdrienCorenflos** [comment]
That should be good now @junpenglao 
Sorry about the hold up @ismael-mendoza but I think this is for a good reason :D

**Review by junpenglao — APPROVED** [review]

**Comment by junpenglao** [comment]
Great stuff, thank you!

**Pull request merged**
